### PR TITLE
Instance verification plugin

### PR DESCRIPTION
**Pull request by ikapelyukhin**
Implements instance verification functionality for RMT. Instance verification modules are provider-specific and are shipped separately.

### CODE REVIEW

**Comment by suse-tests-pass** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          8 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>db/migrate/20190102125933_add_hw_infos_extra.rb</td>
      <td>3</td>
      <td>Layout/IndentationWidth: Use 2 (not 4) spaces for indentation. (https://github.com/bbatsov/ruby-style-guide#spaces-indentation)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/bin/rails</td>
      <td>5</td>
      <td>Style/ExpandPathArguments: Use <code>expand_path('..', __dir__)</code> instead of <code>expand_path('../..', __FILE__)</code>.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/bin/rails</td>
      <td>6</td>
      <td>Style/ExpandPathArguments: Use <code>expand_path('../lib/instance_verification/engine', __dir__)</code> instead of <code>expand_path('../../lib/instance_verification/engine', __FILE__)</code>.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/bin/rails</td>
      <td>10</td>
      <td>Style/ExpandPathArguments: Use <code>expand_path('../Gemfile', __dir__)</code> instead of <code>expand_path('../../Gemfile', __FILE__)</code>.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/lib/instance_verification/providers/example.rb</td>
      <td>5</td>
      <td>Lint/UnusedMethodArgument: Unused method argument - <code>product_hash</code>. If it’s necessary, use <code>_</code> or <code>_product_hash</code> as an argument name to indicate that it won’t be used. You can also write as <code>verify(*)</code> if you want the method to accept any arguments but don’t care about them. (https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/lib/instance_verification/providers/example.rb</td>
      <td>5</td>
      <td>Lint/UnusedMethodArgument: Unused method argument - <code>instance_data</code>. If it’s necessary, use <code>_</code> or <code>_instance_data</code> as an argument name to indicate that it won’t be used. You can also write as <code>verify(*)</code> if you want the method to accept any arguments but don’t care about them. (https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/lib/instance_verification/providers/example.rb</td>
      <td>9</td>
      <td>Layout/TrailingBlankLines: Final newline missing. (https://github.com/bbatsov/ruby-style-guide#newline-eof)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>engines/instance_verification/lib/instance_verification/version.rb</td>
      <td>2</td>
      <td>Style/MutableConstant: Freeze mutable objects assigned to constants.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Unless this is a trivial change, please include a CHANGELOG entry.<br />
Run <code>osc vc</code> in the <code>package</code> directory to add one.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by ikapelyukhin** [comment]
@rjschwei @bear454 OK, this PR should be also good to go.

File: engines/instance_verification/lib/instance_verification.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+module InstanceVerification::Providers
+end
+
+require 'instance_verification/engine'
+require 'instance_verification/provider_base'
+
+providers = Dir.glob(File.join(__dir__, 'instance_verification/providers/*.rb'))
+
+raise 'Too many instance verification providers found' if providers.size > 1
```
Comment (rjschwei) [inline engines/instance_verification/lib/instance_verification.rb]: This forces anyone to implement verification in one file or spread the verification into multiple directories. I am OK with this, we just have to be aware of the restriction and communicate when we re-write the setup documentation.

File: Gemfile [hunk 2]
```diff
@@ -89,3 +89,6 @@ gem 'gettext_i18n_rails'
 
 gem 'config', '~> 1.0'
 gem 'terminal-table', '~> 1.8'
+
+# needed by rmt-server-pubcloud
+gem 'jwt', '~> 2.1'
```
Comment (felixsch) [inline Gemfile]: Is this really required? At least in the changes I could not find anything using `JWT`. Maybe I missed it

File: engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+      context 'when verification provider returns false' do
+        before do
+          expect(InstanceVerification::Providers::Example).to receive(:new)
+            .with(be_a(ActiveSupport::Logger), be_a(ActionDispatch::Request), payload, instance_data).and_return(plugin_double)
+          expect(plugin_double).to receive(:instance_valid?).and_return(false)
+          post url, params: payload, headers: headers
+        end
+
+        let(:plugin_double) { instance_double('InstanceVerification::Providers::Example') }
```
Comment (felixsch) [inline engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb]: nitpicking: Why not move it to the upper level context on remove the other definitions of it?

File: Gemfile (hunk 2 above)
Comment (felixsch) [reply Gemfile]: Okay it's apparently required.

File: engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb (hunk 3 above)
Comment (ikapelyukhin) [reply engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb]: Sure, fixed.

**Pull request merged**